### PR TITLE
Add label broadcasting to `cross_entropy_with_logits` and `weighted_cross_entropy_with_logits`

### LINKED ISSUE #51967: tf.nn.sigmoid_cross_entropy_with_logits should support broadcasting
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): no
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Debian 
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: NA
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.6
- Python version: 3.9
- Bazel version (if compiling from source): NA
- GCC/Compiler version (if compiling from source): NA
- CUDA/cuDNN version: NA
- GPU model and memory: NA

**Describe the current behavior**
`tf.nn.sigmoid_cross_entropy_with_logits` does not support broadcasting. 
This leads to some wrong behavior in certain cases (e.g., keras-team/keras#15247) 
**Describe the expected behavior**
I would expect it to support boradcasting

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): yes
- Briefly describe your candidate solution(if contributing):
Add broadcasting for `labels`

**Standalone code to reproduce the issue**
```python
y = tf.random.uniform((10, 1))

# This one works
tf.keras.losses.binary_crossentropy(0.5, y) 

# This one fails
tf.keras.losses.binary_crossentropy(0.5, y, from_logits=True)
```

### PR DESCRIPTION
**Pull request by eli-osherovich**
Fixes #51967.

### CODE REVIEW

**Comment by allenlavoie** [comment]
What is the use-case for automatic broadcasting here? It does make the scalar case in the bug report look nicer than explicit broadcasting, but it could also allow some very serious silent bugs, e.g. forget to reduce along an axis of `logits` and end up with a garbage loss.

Given that `tf.broadcast_to` exists, it's not hard to add broadcasting for things that do need it. So I think there would need to be significant benefits to offset the potential bugs.

**Comment by eli-osherovich** [comment]
@allenlavoie 

The usecase is pretty obvious, I think
1. Usage convenience
2. No out of the blue errors like in the example I gave above.
3.  API uniformity 

I am not sure I see the dangers of the  bugs you are afraid of. Can you please be more specific?



Can you please be more specific about

**Comment by eli-osherovich** [comment]
@allenlavoie 

The usecase is pretty obvious, I think
1. Usage convenience
2. No out of the blue errors like in the example I gave above.
3.  API uniformity 

I am not sure I see the dangers of the  bugs you are afraid of. Can you please be more specific?



Can you please be more specific about

**Comment by allenlavoie** [comment]
Example issue:

```
multi_head_attention = ...  # [batch size, num_heads, internal]
# Oops, should have reshaped to [batch size, num_heads * internal] first
logits = dense(multi_head_attention)  # -> [batch size, num_heads, num_classes]
loss = tf.nn.cross_entropy_with_logits(labels=...  # [batch size, num_classes], logits=logits)
loss = tf.reduce_sum(loss)
```

If num_classes == num_heads, you silently get nonsense.

I agree that lots of other TF symbols do broadcasting, I'm just trying to figure out whether the benefits to real models outweigh the costs to real models. I don't immediately see a practical use for the single scalar label idea, which is why I was asking about the use-case.

**Comment by eli-osherovich** [comment]
Thanks @allenlavoie 

I sort of anticipated such surprises, therefore I limited the broadcasting in the PR. Instead of general broadcasting of `labels` i only allow them to be be broadcasted to the shape of `logits`. Namely, the results is always the same shape as the shape of `logits`. 

P. S.
Please note that is it not limited to the scalar case only. For example `labels` can be of shape `(m, 1)`, whereas `logits`  `(m, n)`

**Comment by eli-osherovich** [comment]
@allenlavoie 

Sorry for being slow. After reading your example again, I do not think that there will be any broadcasting at all.....
What I meant in my previous post, when I spoke about surprising results due to broadcasting was something like this:

```
a = tf.random.uniform((5, 1))
b = tf.random.uniform((5,))
a + b  # this one has shape  (5,5), not (5,1) !!!
```

**Comment by eli-osherovich** [comment]
@allenlavoie 
Friendly ping.

**Comment by allenlavoie** [comment]
I still don't think adding broadcasting here is a good idea. For general math ops it is a source of bugs but also has quite a bit of utility. I just don't see the utility for these loss functions, and I think the potential bugs outweigh the consistency argument. 

You're free to seek another opinion.

**Comment by eli-osherovich** [comment]
It's a pity. One of the reasons why TF looses its ground to PyTorch in my view.

**Comment by gbaned** [comment]
@eli-osherovich Can you please resolve conflicts? Thanks!

**Comment by eli-osherovich** [comment]
@gbaned 
Are you referring to the conflicting files?
What is the point to fix it if this PR is not gonna be accepted?

**Comment by gbaned** [comment]
@eli-osherovich Can you please resolve conflicts? Thanks!

**Comment by eli-osherovich** [comment]
@gbaned 
Please confirm first that this PR will be merged.

**Comment by gbaned** [comment]
@eli-osherovich Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@eli-osherovich Can you please resolve conflicts? Thanks!

**Comment by mihaimaruseac** [comment]
Closing the PR as it doesn't look like it is something we want to do.

**Comment by eli-osherovich** [comment]
@mihaimaruseac 
Sure. If you hate your users.

**Comment by rohan100jain** [comment]
This is a keras broadcasting semantics question. Lets follow up on the keras bug you opened and they should specify the broadcasting semantics they would like to offer. From there, its an implementation question - keras can easily call tf.broadcast_to and make it uniform before calling TF's low level API which expects not to broadcast (we can make it clear in documentation that this API doesn't do broadcasting)

From TF's perspective, we don't see the benefit of adding broadcasting to a low level API. The potential cons outweigh the pros. TF is a more general purpose library than Keras and we have a broader set of use cases to consider.
